compiler/rustc_mir/src/interpret/intrinsics/gc_layout.rs

```diff
@@ -0,0 +1,72 @@
+use rustc_middle::mir::interpret::Allocation;
+use rustc_middle::ty::layout::LayoutCx;
+use rustc_middle::ty::{ParamEnv, Ty, TyCtxt};
+use rustc_span::symbol::sym;
+use rustc_target::abi::{Align, Size};
+use std::mem::size_of;
+
+fn gc_layout(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, param_env: ParamEnv<'tcx>) -> Vec<u8> {
+    let layout = tcx.layout_of(param_env.and(ty)).unwrap();
+    let w_size = Size::from_bytes(size_of::<usize>());
+    let w_align = Align::from_bytes(w_size.bytes()).unwrap();
+
+    if layout.align.abi.bytes() < (w_size.bytes()) {
+        // If an ADT is aligned to a size smaller than 1 word, then it
+        // contains no fields large enough to represent a GC pointer.
+        //
+        // By default, the collector will conservatively scan a struct's
+        // memory, treating it as word-sized chunks. However, since rust is
+        // free to shuffle field order around, any in-memory words which
+        // represent actual pointers would be purely accidental. Such
+        // structs should be ignored completely by the collector.
+        return vec![0; 16]; // empty (bitmap, size) (u64,u64) as byte stream
+    }
+
+    let src_field_iter = ty.ty_adt_def().unwrap().all_fields();
+    let mut bitmap: u64 = 0xFFFFFFFF;
+
+    for i in layout.fields.index_by_increasing_offset() {
+        let target_offset = layout.fields.offset(i as usize);
+
+        let layout_cx = LayoutCx { tcx, param_env };
+        let field_size = layout.field(&layout_cx, i).unwrap().size;
+
+        if field_size < w_size {
+            // If a field is smaller than a word, it will either be shuffled
+            // around in the struct or padded to the min alignment with
+            // uninitialized memory. Whatever the case, the word it lives in is
+            // never going to be a valid pointer.
+            let mask = target_offset.bytes() / w_align.bytes();
```

Comment:
I'm not sure how that makes the correct mask. Does it need commentary?

Reply:
`target_offset` is the byte in the struct's layout where this field begins. If we divide it by `w_align` (8 bytes) we get the offset in words. We then mask out the bit in the bitmap that corresponds to that fields word offset.

Reply:
Does that rather good explanation need to be a comment?

Reply:
Sure, how about 43d49a3836e3f588da0e25e777e7b9a6240944c7?